### ISSUE TITLE
False positive on moved out of ref diagnostic

### ISSUE THREAD

**Issue by ronnodas** [rust-lang/rust-analyzer#15787]
The following code compiles fine ([playground link](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=adf6b8cac7aca66f40c906468b4e5fa9)) but rust-analyzer thinks it's moving a `[u32]` out of a reference (E0507):
```rust
fn foo(mut slice: &[u32]) -> usize {
    slice = match slice {
        [0, rest @ ..] | rest => rest,
    };
    slice.len()
}
```

**rust-analyzer version**:  0.3.1697-standalone

**rustc version**: rustc 1.73.0 (cc66ad468 2023-10-03)

**relevant settings**: `"rust-analyzer.diagnostics.enable": true, "rust-analyzer.diagnostics.experimental.enable": true`

**Comment by HKalbasi** [comment]
`"rust-analyzer.diagnostics.experimental.enable": false` Is this correct?

**Comment by lnicola** [comment]
I can reproduce it only with experimental diagnostics enabled.

**Comment by ronnodas** [comment]
Sorry, I tried with experimental diagnostics both enabled and disabled and maybe it was some kind of caching or settings syncing issue. I too can only reproduce it with that setting enabled now, so edited the original post.
